### ISSUE TITLE
[RFC]: Decoupling Text Encoder in Diffusion Model Inference Pipeline

### ISSUE THREAD

**Issue by ZJY0516** [vllm-project/vllm-omni#771]
### Motivation.

Diffusion Model Current Inference Flow
```python
pre_process_func (if any)
add_request() -> pipeline.forward()
return result
```

`pipeline.forward()` steps:

- Check inputs
- Encode prompts
- Prepare latents and timesteps
- Diffusion process
- Decode latents

Diffusion Components are in the same process
- text_encoder
- tokenizer
- transformer
- scheduler
- vae

For performance and engineering effort considerations, we directly reused the transformers for text_encoder. However, the current architecture has some issues:

- During Tensor Parallelism (TP), each TP process maintains a full copy of the text_encoder weights (due to Transformers library's lack of TP support)

- Component initialization is tightly coupled: pipeline initialization loads all weights simultaneously, complicating weight offloading strategies- 

### Proposed Change.

Decouple the text_encoder into a separate process to prevent OOM during GPU memory constraints.

Implementation Options
1. **A New Worker in Diffusion Engine**
Design a separate worker specifically for text_encoder within the diffusion engine
Advantage: Simpler implementation

2. **Stage-based Separation**
Treat text_encoder as Stage 1
Dit and vae components as Stage 2
Advantage: Aligns with overall project architecture design

**Goal**

- Reduce GPU memory pressure during inference
- Enable more flexible weight management
- Improve scalability of the diffusion pipeline 

**Note**

This approach primarily targets models with lightweight text encoders. For models like Bagel, which feature computationally intensive text encoders, we naturally leverage the vLLM engine and the stage abstraction of vLLM-Omni.

---

**update in 2026.01.27**

vllm can return hidden states of last layer, and it's very fast. But vllm shows an accuracy gap relative to transformers

<details>
<summary>benchmark code</code></summary>

```python
import random
import time
from argparse import Namespace

import torch
from transformers import AutoModel
from vllm import LLM, EngineArgs, TokensPrompt
from vllm.pooling_params import PoolingParams
from vllm.utils.argparse_utils import FlexibleArgumentParser

MODEL_NAME = "Qwen/Qwen2-7B"
# Qwen/Qwen2.5-VL-3B-Instruct
SEQ_LEN = 512


# -------------------------
# Utils
# -------------------------
def random_token_prompt(seq_len: int, base_min=1024, base_max=10240):
    base = random.randint(base_min, base_max)
    return [base + i for i in range(seq_len)]


def benchmark(fn, num_run: int):
    start = time.perf_counter()
    for _ in range(num_run):
        fn()
    torch.cuda.synchronize()
    return (time.perf_counter() - start) / num_run


# -------------------------
# HF & vLLM wrappers
# -------------------------
def run_hf_forward(model, token_ids):
    output = model(
        input_ids=torch.tensor([token_ids]),
        output_hidden_states=True,
    )
    return output.hidden_states[-1].squeeze(0)


def run_vllm_encode(llm, token_ids, pooling_params):
    output = llm.encode(
        TokensPrompt(prompt_token_ids=token_ids),
        pooling_task="token_embed",
        pooling_params=pooling_params,
        use_tqdm=False,
    )
    return output[0].outputs.data


# -------------------------
# Args
# -------------------------
def parse_args():
    parser = FlexibleArgumentParser()
    parser = EngineArgs.add_cli_args(parser)
    parser.set_defaults(
        model=MODEL_NAME,
        runner="pooling",
        enable_prefix_caching=False,
    )
    return parser.parse_args()


# -------------------------
# Main
# -------------------------
def main(args: Namespace):
    llm = LLM(**vars(args))
    hf_model = AutoModel.from_pretrained(MODEL_NAME, trust_remote_code=True)

    pooling_params = PoolingParams(use_activation=False)

    # correctness check
    token_ids = random_token_prompt(SEQ_LEN, base_min=1024, base_max=5120)

    hf_hidden = run_hf_forward(hf_model, token_ids)
    vllm_hidden = run_vllm_encode(llm, token_ids, pooling_params)

    print("Max diff:", (hf_hidden - vllm_hidden).abs().max().item())

    # performance benchmark
    num_run = 10

    vllm_time = benchmark(
        lambda: run_vllm_encode(
            llm,
            random_token_prompt(SEQ_LEN),
            pooling_params,
        ),
        num_run,
    )

    hf_time = benchmark(
        lambda: run_hf_forward(
            hf_model,
            random_token_prompt(SEQ_LEN),
        ),
        num_run,
    )

    print(f"vLLM Time per run: {vllm_time:.6f} seconds")
    print(f"HuggingFace Time per run: {hf_time:.6f} seconds")


if __name__ == "__main__":
    args = parse_args()
    with torch.inference_mode():
        main(args)

```

</details>

```
Max diff: 69.94607543945312
vLLM Time per run: 0.059688 seconds
HuggingFace Time per run: 2.024002 seconds
```

**My current plan is that we provide 2 backends for users in text_encoder stage, one is vllm, and the other is transformers/encoders in diffusers**

Note: some encoders don't has beed implemented in vLLM, e.g. UMT5EncoderModel used in WAN 2.2. And some models has more than 1 text encoders, we will not support use vLLM in this case

### CC List.

@hsliuustc0106 @SamitHuang @fake0fan @Gaohan123 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by david6666666** [comment]
Will the newly added text encoder worker be placed under vllm_omni/diffusion/worker, or will the Omni model and diffusion model workers be merged later?

**Comment by ZJY0516** [comment]
The worker/engine of text_encoder will be very simple, just a process to run transformers, receive prompts(may including image) and return embeddings. 

I think no need to merge it to any existing worker/engine now

**Comment by david6666666** [comment]
> The worker/engine of text_encoder will be very simple, just a process to run transformers, receive prompts(may including image) and return embeddings.
> 
> I think no need to merge it to any existing worker/engine now

makes sense, looking forward to the follow-up PR.

**Comment by hsliuustc0106** [comment]
this should also be useful for continuous batch since we can decouple preprocessing with the rest part

**Comment by hsliuustc0106** [comment]
<img width="583" height="319" alt="Image" src="https://github.com/user-attachments/assets/547807cf-7e00-4ec4-b256-8503213bd0e4" /> @asukaqaq-s

**Comment by fake0fan** [comment]
The overall design is reasonable, I'm just curious about the typical model size for the text encoder part.

**Comment by ZJY0516** [comment]
overall, the text_encoder is usually not very large compared to the full model — often about 1/3

- qwen-image 7B qwen2.5 vl
- z-image 4B qwen3 vl
- flux2.dev 24B mistral3

**Comment by hsliuustc0106** [comment]
> overall, the text_encoder is usually not very large compared to the full model — often about 1/3
> 
> * qwen-image 7B qwen2.5 vl
> * z-image 4B qwen3 vl
> * flux2.dev 24B mistral3

are we going to use vllm AR to accelerate?

**Comment by ZJY0516** [comment]
> are we going to use vllm AR to accelerate?

For models with lightweight text encoders, the answer is no. For models like Bagel amd glm-image, which feature computationally intensive text encoders, we naturally leverage the vLLM engine and the stage abstraction of vLLM-Omni.

**Comment by hsliuustc0106** [comment]
> > are we going to use vllm AR to accelerate?
> 
> For models with lightweight text encoders, the answer is no. For models like Bagel amd glm-image, which feature computationally intensive text encoders, we naturally leverage the vLLM engine and the stage abstraction of vLLM-Omni.

make sense

**Comment by asukaqaq-s** [comment]
> this should also be useful for continuous batch since we can decouple preprocessing with the rest part



> <img alt="Image" width="583" height="319" src="https://private-user-images.githubusercontent.com/222337142/535091353-547807cf-7e00-4ec4-b256-8503213bd0e4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgzNzI4MDMsIm5iZiI6MTc2ODM3MjUwMywicGF0aCI6Ii8yMjIzMzcxNDIvNTM1MDkxMzUzLTU0NzgwN2NmLTdlMDAtNGVjNC1iMjU2LTg1MDMyMTNiZDBlNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExNFQwNjM1MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YWQ5NmU5ZWFmNjE3MTA0ZmE4NjUxNDljOTM4ZWE3MDZjM2JkMDU0MmE0MjRlOTAzZDIyMmM1MTU1YjM4NDZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.uGwumJTXMY1bhdYQraW-N0fBnoEHaNUEf0QwSYkd8ok"> [@asukaqaq-s](https://github.com/asukaqaq-s)

Fully agree.
This decouples VAE and text encoder preprocessing from the denoising path, similar to how vLLM handles image/audio encoders, and is key for step-level continuous batching.

**Comment by pengchengneo** [comment]
gread work, and I wonder will vae component will be decoupled from Stage2 in the future @ZJY0516

**Comment by hsliuustc0106** [comment]
> gread work, and I wonder will vae component will be decoupled from Stage2 in the future [@ZJY0516](https://github.com/ZJY0516)

this should be next step, right? @ZJY0516

**Comment by ZJY0516** [comment]
> > gread work, and I wonder will vae component will be decoupled from Stage2 in the future [@ZJY0516](https://github.com/ZJY0516)
> 
> this should be next step, right? [@ZJY0516](https://github.com/ZJY0516)

I have discussed this with @SamitHuang offline. We are not sure about this

**Comment by ZJY0516** [comment]
FYI, vllm can return hidden states of last layer, and it's very fast. But vllm shows an accuracy gap relative to transformers

<details>
<summary>benchmark code</code></summary>

```python
import random
import time
from argparse import Namespace

import torch
from transformers import AutoModel
from vllm import LLM, EngineArgs, TokensPrompt
from vllm.pooling_params import PoolingParams
from vllm.utils.argparse_utils import FlexibleArgumentParser

MODEL_NAME = "Qwen/Qwen2-7B"
# Qwen/Qwen2.5-VL-3B-Instruct
SEQ_LEN = 512


# -------------------------
# Utils
# -------------------------
def random_token_prompt(seq_len: int, base_min=1024, base_max=10240):
    base = random.randint(base_min, base_max)
    return [base + i for i in range(seq_len)]


def benchmark(fn, num_run: int):
    start = time.perf_counter()
    for _ in range(num_run):
        fn()
    torch.cuda.synchronize()
    return (time.perf_counter() - start) / num_run


# -------------------------
# HF & vLLM wrappers
# -------------------------
def run_hf_forward(model, token_ids):
    output = model(
        input_ids=torch.tensor([token_ids]),
        output_hidden_states=True,
    )
    return output.hidden_states[-1].squeeze(0)


def run_vllm_encode(llm, token_ids, pooling_params):
    output = llm.encode(
        TokensPrompt(prompt_token_ids=token_ids),
        pooling_task="token_embed",
        pooling_params=pooling_params,
        use_tqdm=False,
    )
    return output[0].outputs.data


# -------------------------
# Args
# -------------------------
def parse_args():
    parser = FlexibleArgumentParser()
    parser = EngineArgs.add_cli_args(parser)
    parser.set_defaults(
        model=MODEL_NAME,
        runner="pooling",
        enable_prefix_caching=False,
    )
    return parser.parse_args()


# -------------------------
# Main
# -------------------------
def main(args: Namespace):
    llm = LLM(**vars(args))
    hf_model = AutoModel.from_pretrained(MODEL_NAME, trust_remote_code=True)

    pooling_params = PoolingParams(use_activation=False)

    # correctness check
    token_ids = random_token_prompt(SEQ_LEN, base_min=1024, base_max=5120)

    hf_hidden = run_hf_forward(hf_model, token_ids)
    vllm_hidden = run_vllm_encode(llm, token_ids, pooling_params)

    print("Max diff:", (hf_hidden - vllm_hidden).abs().max().item())

    # performance benchmark
    num_run = 10

    vllm_time = benchmark(
        lambda: run_vllm_encode(
            llm,
            random_token_prompt(SEQ_LEN),
            pooling_params,
        ),
        num_run,
    )

    hf_time = benchmark(
        lambda: run_hf_forward(
            hf_model,
            random_token_prompt(SEQ_LEN),
        ),
        num_run,
    )

    print(f"vLLM Time per run: {vllm_time:.6f} seconds")
    print(f"HuggingFace Time per run: {hf_time:.6f} seconds")


if __name__ == "__main__":
    args = parse_args()
    with torch.inference_mode():
        main(args)

```

</details>

```
Max diff: 69.94607543945312
vLLM Time per run: 0.059688 seconds
HuggingFace Time per run: 2.024002 seconds
```

**My current plan is that we provide 2 backends for users in text_encoder stage, one is vllm, and the other is transformers/encoders in diffusers**

Note: some encoders don't has beed implemented in vLLM, e.g. UMT5EncoderModel used in WAN 2.2. And some models has more than 1 text encoders, we will not support use vLLM in this case

**Comment by hsliuustc0106** [comment]
> FYI, vllm can return hidden states of last layer, and it's very fast. But vllm shows an accuracy gap relative to transformers
> 
> benchmark code
> import random
> import time
> from argparse import Namespace
> 
> import torch
> from transformers import AutoModel
> from vllm import LLM, EngineArgs, TokensPrompt
> from vllm.pooling_params import PoolingParams
> from vllm.utils.argparse_utils import FlexibleArgumentParser
> 
> MODEL_NAME = "Qwen/Qwen2-7B"
> # Qwen/Qwen2.5-VL-3B-Instruct
> SEQ_LEN = 512
> 
> 
> # -------------------------
> # Utils
> # -------------------------
> def random_token_prompt(seq_len: int, base_min=1024, base_max=10240):
>     base = random.randint(base_min, base_max)
>     return [base + i for i in range(seq_len)]
> 
> 
> def benchmark(fn, num_run: int):
>     start = time.perf_counter()
>     for _ in range(num_run):
>         fn()
>     torch.cuda.synchronize()
>     return (time.perf_counter() - start) / num_run
> 
> 
> # -------------------------
> # HF & vLLM wrappers
> # -------------------------
> def run_hf_forward(model, token_ids):
>     output = model(
>         input_ids=torch.tensor([token_ids]),
>         output_hidden_states=True,
>     )
>     return output.hidden_states[-1].squeeze(0)
> 
> 
> def run_vllm_encode(llm, token_ids, pooling_params):
>     output = llm.encode(
>         TokensPrompt(prompt_token_ids=token_ids),
>         pooling_task="token_embed",
>         pooling_params=pooling_params,
>         use_tqdm=False,
>     )
>     return output[0].outputs.data
> 
> 
> # -------------------------
> # Args
> # -------------------------
> def parse_args():
>     parser = FlexibleArgumentParser()
>     parser = EngineArgs.add_cli_args(parser)
>     parser.set_defaults(
>         model=MODEL_NAME,
>         runner="pooling",
>         enable_prefix_caching=False,
>     )
>     return parser.parse_args()
> 
> 
> # -------------------------
> # Main
> # -------------------------
> def main(args: Namespace):
>     llm = LLM(**vars(args))
>     hf_model = AutoModel.from_pretrained(MODEL_NAME, trust_remote_code=True)
> 
>     pooling_params = PoolingParams(use_activation=False)
> 
>     # correctness check
>     token_ids = random_token_prompt(SEQ_LEN, base_min=1024, base_max=5120)
> 
>     hf_hidden = run_hf_forward(hf_model, token_ids)
>     vllm_hidden = run_vllm_encode(llm, token_ids, pooling_params)
> 
>     print("Max diff:", (hf_hidden - vllm_hidden).abs().max().item())
> 
>     # performance benchmark
>     num_run = 10
> 
>     vllm_time = benchmark(
>         lambda: run_vllm_encode(
>             llm,
>             random_token_prompt(SEQ_LEN),
>             pooling_params,
>         ),
>         num_run,
>     )
> 
>     hf_time = benchmark(
>         lambda: run_hf_forward(
>             hf_model,
>             random_token_prompt(SEQ_LEN),
>         ),
>         num_run,
>     )
> 
>     print(f"vLLM Time per run: {vllm_time:.6f} seconds")
>     print(f"HuggingFace Time per run: {hf_time:.6f} seconds")
> 
> 
> if __name__ == "__main__":
>     args = parse_args()
>     with torch.inference_mode():
>         main(args)
> ```
> Max diff: 69.94607543945312
> vLLM Time per run: 0.059688 seconds
> HuggingFace Time per run: 2.024002 seconds
> ```
> 
> **My current plan is that we provide 2 backends for users in text_encoder stage, one is vllm, and the other is transformers/encoders in diffusers**
> 
> Note: some encoders don't has beed implemented in vLLM, e.g. UMT5EncoderModel used in WAN 2.2. And some models has more than 1 text encoders, we will not support use vLLM in this case

Agree, the accuracy of diffusion models are hard to control

**Comment by tzhouam** [comment]
@ZJY0516 From the script, seems like the vllm is on bf16 by default. Does this matter?
